File: packages/auths-python/python/auths/identity.py
"""Identity and agent resource services — Stripe-style API."""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING

from auths._native import (
    create_identity as _create_identity,
    create_agent_identity as _create_agent_identity,
    delegate_agent as _delegate_agent,
    rotate_identity_ffi as _rotate_identity,
)
from auths.rotation import IdentityRotationResult

if TYPE_CHECKING:
    from auths._client import Auths


@dataclass
class Identity:
    """An Auths identity (represents a did:keri: identifier)."""

    did: str
    key_alias: str
    label: str
    repo_path: str
    public_key: str


@dataclass
class AgentIdentity:
    """Standalone agent identity (did:keri:). Created via identities.create_agent()."""

    did: str
    key_alias: str
    attestation: str
    public_key: str


@dataclass
class DelegatedAgent:
    """Agent delegated under a parent identity (did:key:). Created via identities.delegate_agent()."""

    did: str
    key_alias: str
    attestation: str
    public_key: str


class IdentityService:
    """Resource service for identity operations.

    Usage:
        auths = Auths()
        identity = auths.identities.create(label="laptop")
        agent = auths.identities.delegate_agent(identity.did, name="ci-bot", capabilities=["sign"])
    """

    def __init__(self, client: Auths):
        self._client = client

    def create(
        self,
        label: str = "main",
        repo_path: str | None = None,
        passphrase: str | None = None,
    ) -> Identity:
        """Create a new identity.

        Args:
            label: Human-readable label for this identity (default: "main").
            repo_path: Git repo path (default: client's repo_path).
            passphrase: Key passphrase (default: client's passphrase or AUTHS_PASSPHRASE env var).

        Usage:
            identity = auths.identities.create(label="laptop")
        """
        rp = repo_path or self._client.repo_path
        pp = passphrase or self._client._passphrase
        did, key_alias, public_key_hex = _create_identity(label, rp, pp)
        return Identity(did=did, key_alias=key_alias, label=label, repo_path=rp, public_key=public_key_hex)

    def rotate(
        self,
        identity_did: str,
        *,
        passphrase: str | None = None,
    ) -> IdentityRotationResult:
        """Rotate an identity's keys using the KERI pre-rotation ceremony.

        This is a single atomic operation. If any step fails, the previous key
        remains active and no partial state is written.

        After rotation:
        - Old attestations remain valid (verified via Key Event Log history)
        - New signing operations use the rotated key automatically
        - Device links are unaffected (bound to DID, not key)

        Args:
            identity_did: The KERI DID of the identity to rotate.
            passphrase: Optional passphrase for keychain access.

        Usage:
            result = auths.identities.rotate(identity.did)
            print(f"Rotated to sequence {result.sequence}")
        """
        pp = passphrase or self._client._passphrase
        native_result = _rotate_identity(self._client.repo_path, None, None, pp)
        return IdentityRotationResult(
            controller_did=native_result.controller_did,
            new_key_fingerprint=native_result.new_key_fingerprint,
            previous_key_fingerprint=native_result.previous_key_fingerprint,
            sequence=native_result.sequence,
        )

    def create_agent(
        self,
        name: str,
        capabilities: list[str],
        passphrase: str | None = None,
    ) -> AgentIdentity:
        """Create a standalone agent identity (did:keri:).

        Args:
            name: Human-readable agent name.
            capabilities: List of capabilities (e.g., ["sign", "verify"]).
            passphrase: Key passphrase override.

        Usage:
            agent = auths.identities.create_agent("ci-bot", ["sign"])
        """
        pp = passphrase or self._client._passphrase
        bundle = _create_agent_identity(
            name, capabilities, self._client.repo_path, pp,
        )
        return AgentIdentity(
            did=bundle.agent_did, key_alias=bundle.key_alias,
            attestation=bundle.attestation_json, public_key=bundle.public_key_hex,
        )

    def delegate_agent(
        self,
        identity_did: str,
        name: str,
        capabilities: list[str],
        expires_in_days: int | None = None,
        passphrase: str | None = None,
    ) -> DelegatedAgent:
        """Delegate an agent under an identity (did:key:).

        Args:
            identity_did: The parent identity's DID.
            name: Human-readable agent name.
            capabilities: List of capabilities (e.g., ["sign", "verify"]).
            expires_in_days: Optional TTL in days.
            passphrase: Key passphrase override.

        Usage:
            agent = auths.identities.delegate_agent(identity.did, "ci-bot", ["sign"])
        """
        pp = passphrase or self._client._passphrase
        bundle = _delegate_agent(
            name, capabilities, self._client.repo_path, pp, expires_in_days,
            identity_did,
        )
        return DelegatedAgent(
            did=bundle.agent_did, key_alias=bundle.key_alias,
            attestation=bundle.attestation_json, public_key=bundle.public_key_hex,
        )

File: packages/auths-node/lib/identity.ts
import native from './native'
import { mapNativeError, CryptoError, IdentityError } from './errors'
import type { Auths } from './client'

/** A cryptographic identity anchored in a KERI key event log. */
export interface Identity {
  /** The KERI decentralized identifier (e.g. `did:keri:EBfd...`). */
  did: string
  /** Keychain alias used to retrieve the signing key. */
  keyAlias: string
  /** Human-readable label for this identity. */
  label: string
  /** Path to the Git registry that stores this identity. */
  repoPath: string
  /** Hex-encoded Ed25519 public key. */
  publicKey: string
}

/** A standalone agent identity with its self-signed attestation. */
export interface AgentIdentity {
  /** The agent's KERI decentralized identifier. */
  did: string
  /** Keychain alias for the agent's signing key. */
  keyAlias: string
  /** JSON-serialized self-signed attestation. */
  attestation: string
  /** Hex-encoded Ed25519 public key. */
  publicKey: string
}

/** An agent delegated under an existing identity. */
export interface DelegatedAgent {
  /** The delegated agent's DID (typically `did:key:z...`). */
  did: string
  /** Keychain alias for the agent's signing key. */
  keyAlias: string
  /** JSON-serialized delegation attestation signed by the parent identity. */
  attestation: string
  /** Hex-encoded Ed25519 public key. */
  publicKey: string
}

/** Result of a key rotation operation. */
export interface RotationResult {
  /** The controller DID whose keys were rotated. */
  controllerDid: string
  /** Fingerprint of the new signing key. */
  newKeyFingerprint: string
  /** Fingerprint of the previous signing key. */
  previousKeyFingerprint: string
  /** New KERI event sequence number after rotation. */
  sequence: number
}

/** Options for {@link IdentityService.create}. */
export interface CreateIdentityOptions {
  /** Human-readable label. Defaults to `'main'`. */
  label?: string
  /** Override the client's repo path. */
  repoPath?: string
  /** Override the client's passphrase. */
  passphrase?: string
}

/** Options for {@link IdentityService.createAgent}. */
export interface CreateAgentOptions {
  /** Name for the agent identity. */
  name: string
  /** Capabilities to grant (e.g. `['sign']`). */
  capabilities: string[]
  /** Override the client's passphrase. */
  passphrase?: string
}

/** Options for {@link IdentityService.delegateAgent}. */
export interface DelegateAgentOptions {
  /** DID of the parent identity that delegates authority. */
  identityDid: string
  /** Name for the delegated agent. */
  name: string
  /** Capabilities to grant (e.g. `['sign']`). */
  capabilities: string[]
  /** Optional expiration in days. */
  expiresInDays?: number
  /** Override the client's passphrase. */
  passphrase?: string
}

/** Options for {@link IdentityService.rotate}. */
export interface RotateKeysOptions {
  /** DID of the identity to rotate. Defaults to the primary identity. */
  identityDid?: string
  /** Override the client's passphrase. */
  passphrase?: string
}

/** Options for {@link IdentityService.getPublicKey}. */
export interface GetPublicKeyOptions {
  /** DID of the identity whose public key to retrieve. */
  identityDid: string
  /** Override the client's passphrase. */
  passphrase?: string
}

/**
 * Manages cryptographic identities, agents, and key rotation.
 *
 * Access via {@link Auths.identities}.
 *
 * @example
 * ```typescript
 * const auths = new Auths()
 * const identity = auths.identities.create({ label: 'laptop' })
 * console.log(identity.did) // did:keri:EBfd...
 * ```
 */
export class IdentityService {
  constructor(private client: Auths) {}

  /**
   * Creates a new cryptographic identity backed by an Ed25519 keypair.
   *
   * @param opts - Creation options.
   * @returns The newly created identity.
   * @throws {@link IdentityError} if the identity cannot be created.
   *
   * @example
   * ```typescript
   * const identity = auths.identities.create({ label: 'laptop' })
   * console.log(identity.did)       // did:keri:EBfd...
   * console.log(identity.publicKey) // hex-encoded Ed25519 key
   * ```
   */
  create(opts: CreateIdentityOptions = {}): Identity {
    const rp = opts.repoPath ?? this.client.repoPath
    const pp = opts.passphrase ?? this.client.passphrase
    try {
      const result = native.createIdentity(opts.label ?? 'main', rp, pp)
      return {
        did: result.did,
        keyAlias: result.keyAlias,
        label: opts.label ?? 'main',
        repoPath: rp,
        publicKey: result.publicKeyHex,
      }
    } catch (err) {
      throw mapNativeError(err, IdentityError)
    }
  }

  /**
   * Creates a standalone agent identity with a self-signed attestation.
   *
   * @param opts - Agent creation options.
   * @returns The agent identity with its attestation.
   * @throws {@link IdentityError} if the agent cannot be created.
   *
   * @example
   * ```typescript
   * const agent = auths.identities.createAgent({
   *   name: 'ci-bot',
   *   capabilities: ['sign'],
   * })
   * console.log(agent.did) // did:keri:...
   * ```
   */
  createAgent(opts: CreateAgentOptions): AgentIdentity {
    const pp = opts.passphrase ?? this.client.passphrase
    try {
      const bundle = native.createAgentIdentity(
        opts.name,
        opts.capabilities,
        this.client.repoPath,
        pp,
      )
      return {
        did: bundle.agentDid,
        keyAlias: bundle.keyAlias,
        attestation: bundle.attestationJson,
        publicKey: bundle.publicKeyHex,
      }
    } catch (err) {
      throw mapNativeError(err, IdentityError)
    }
  }

  /**
   * Delegates an agent under an existing identity with scoped capabilities.
   *
   * @param opts - Delegation options.
   * @returns The delegated agent with its signed attestation.
   * @throws {@link IdentityError} if delegation fails.
   *
   * @example
   * ```typescript
   * const agent = auths.identities.delegateAgent({
   *   identityDid: identity.did,
   *   name: 'deploy-bot',
   *   capabilities: ['sign'],
   *   expiresInDays: 90,
   * })
   * ```
   */
  delegateAgent(opts: DelegateAgentOptions): DelegatedAgent {
    const pp = opts.passphrase ?? this.client.passphrase
    try {
      const bundle = native.delegateAgent(
        opts.name,
        opts.capabilities,
        this.client.repoPath,
        pp,
        opts.expiresInDays ?? null,
        opts.identityDid,
      )
      return {
        did: bundle.agentDid,
        keyAlias: bundle.keyAlias,
        attestation: bundle.attestationJson,
        publicKey: bundle.publicKeyHex,
      }
    } catch (err) {
      throw mapNativeError(err, IdentityError)
    }
  }

  /**
   * Rotates the signing keys for an identity, advancing the KERI event log.
   *
   * @param opts - Rotation options.
   * @returns The rotation result with old and new key fingerprints.
   * @throws {@link IdentityError} if rotation fails.
   *
   * @example
   * ```typescript
   * const result = auths.identities.rotate({ identityDid: identity.did })
   * console.log(result.sequence) // incremented sequence number
   * ```
   */
  rotate(opts: RotateKeysOptions = {}): RotationResult {
    const pp = opts.passphrase ?? this.client.passphrase
    try {
      const result = native.rotateIdentityKeys(
        this.client.repoPath,
        opts.identityDid ?? null,
        null,
        pp,
      )
      return {
        controllerDid: result.controllerDid,
        newKeyFingerprint: result.newKeyFingerprint,
        previousKeyFingerprint: result.previousKeyFingerprint,
        sequence: result.sequence,
      }
    } catch (err) {
      throw mapNativeError(err, IdentityError)
    }
  }

  /**
   * Retrieves the hex-encoded Ed25519 public key for an identity.
   *
   * @param opts - Lookup options.
   * @returns Hex-encoded public key string (64 characters).
   * @throws {@link CryptoError} if the key cannot be found.
   *
   * @example
   * ```typescript
   * const pk = auths.identities.getPublicKey({ identityDid: identity.did })
   * console.log(pk.length) // 64
   * ```
   */
  getPublicKey(opts: GetPublicKeyOptions): string {
    const pp = opts.passphrase ?? this.client.passphrase
    try {
      return native.getIdentityPublicKey(opts.identityDid, this.client.repoPath, pp)
    } catch (err) {
      throw mapNativeError(err, CryptoError)
    }
  }
}
